trace2: add infrastructure and initial events

ldennington opened the pull request:
Add initial `TRACE2` functionality, including:

1. Small refactorings to make `TryGetAssemblyVersion` accessible
outside of the `DiagnosticCommand` class and shared trace logic
accessible from a `TraceUtils` class.
2. Addition of a `Trace2CollectorWriter` class, which handles writing
to the Telemetry Tool/OTel collector and a `Trace2StreamWriter` class,
which handles writing to stderr/files.
3. Basic `TRACE2` functionality, including the ability to add writers to
different format targets, logic to send messages to these writers, and
the ability to release these writers before application exit.
4. Support for session IDs.
5. Ability for users to enable normal/event format targets.
6. Writing `Version`, `Start`, and `Exit` events.

While this patch series does not represent all the `TRACE2` events we plan
to implement, it does provide the opportunity to review and suggest changes
to the foundation these events are built on. It also provides the general pattern
that additional patches will use to implement new events, just in case that
needs to be tweaked.

derrickstolee commented on src/shared/Core/CollectorWriter.cs:
```diff
@@ -0,0 +1,172 @@
+    /// Accepts string messages from multiple threads and dispatches them over a named pipe from a
+    /// background thread.
+    /// </summary>
+    public class CollectorWriter
```
Does this need an explicit `: IDisposable`?

derrickstolee replied:
Or, as I see later, `DisposableObject`?

ldennington replied:
Yep, it absolutely does. Good catch.

derrickstolee commented on src/shared/Core/CollectorWriter.cs:
```diff
@@ -0,0 +1,172 @@
+                // background thread loop if the pipe is not available.
+                pipe.Connect(timeout: 0);
+
+                _pipeClient = pipe;
```
This commit makes me need to page-in all of the rules around `IDisposable` and this late copy-to-member is one of those things. Makes sense that we don't want to call `_pipeClient.Dispose()` if it failed to `Connect()` in the line above.

jeffhostetler replied:
I never have understood the rules of IDisposable in C#, so I'll defer to your freshly memory mapped pages.

derrickstolee replied:
Thinking about it a day later, I'm not sure if I was right. The client's `Dispose()` should be resilient to a failed setup, so this precaution isn't necessary. In fact, the current code will not `Dispose()` the client if it fails to connect, which is not the correct thing to do. Perhaps we should be acting on `_pipeClient` and drop the local `pipe` variable?

ldennington replied:
I'll admit to being a bit confused here. Is the consensus that we should be calling `Dispose` if the `Connect` call fails?

derrickstolee replied:
I think the ideas is that you shouldn't have a method-local `pipe` variable and instead act on `_pipeClient` the entire time.

derrickstolee commented on src/shared/Core/Trace2.cs:
```diff
@@ -0,0 +1,93 @@
+/// </summary>
+public interface ITrace2 : IDisposable
+{
+    /// <summary>
+    /// Add a listener to the trace writer.
+    /// </summary>
+    /// <param name="listener">The listener to add.</param>
+    void AddListener(CollectorWriter listener);
```
Ok, as of this commit, `ITrace2` objects _emit_ events, but do not _accept_ events. I suppose that makes sense for the purposes of start/stop events. I expect you'll update it when adding other kinds of logging.

jeffhostetler replied:
Or are we using "listener" in the context of a background queue thread that is receiving trace messages from the foreground thread and that it will relay trace events to the daemon (aka collector service) ?

ldennington replied:
Yes, the `CollectorWriter` that is added here sets up the queue that receives TRACE2 messages and eventually writes them to the daemon. Perhaps `AddWriter` would be a more appropriate name...

derrickstolee commented on src/shared/Core/Constants.cs:
```diff
@@ -49,6 +49,7 @@ public static class EnvironmentVariables
             public const string GcmTrace              = "GCM_TRACE";
             public const string GcmTraceSecrets       = "GCM_TRACE_SECRETS";
             public const string GcmTraceMsAuth        = "GCM_TRACE_MSAUTH";
+            public const string GcmTrace2             = "GCM_TRACE2";
```
A big picture question: Are we always planning to have a separate `GCM_TRACE2` variable, or will we be piggy-backing `GIT_TRACE2*` variables eventually?

jeffhostetler replied:
Yeah, I'm wondering about whether we should read a GCM_TRACE2 or the GIT_TRACE2_EVENT and/or `git config trace2.eventtarget`.

ldennington replied:
I think it probably does make more sense to piggyback here - my initial thinking was to give users a bit more control over whether trace2 information is sent, but it could prove to be a hassle to enable for both Git and GCM.

mjcheetham replied:
+1 for using just `GIT_` vars.

derrickstolee commented on src/shared/Core/Trace2.cs:
```diff
@@ -69,6 +70,22 @@ public interface ITrace2 : IDisposable
+    /// <param name="filePath">Path of the file this method is called from.</param>
+    /// <param name="lineNumber">Line number of file this method is called from.</param>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1026:DefaultParametersShouldNotBeUsed")]
+    void WriteExit(
```
Are we calling this from anywhere? Do we have an existing handler for process termination?

derrickstolee replied:
I suppose it could be handled within the `Dispose()` method of the `Trace2` class, but that's probably not the best place.

ldennington replied:
Sigh. I took a stab at implementing this (with a comment along the lines of `// TODO: Determine whether this is the right place to record this.`) but it appears to have fallen off during one of the many rebases I did to try and get this PR in shape. Thanks for pointing this out!

derrickstolee reviewed (commented):
Thanks for helping me refresh my C#. Lots of good stuff in here.

jeffhostetler commented on src/shared/Core.Tests/SettingsTests.cs:
```diff
@@ -280,6 +279,78 @@ public void Settings_IsTracingEnabled_EnvarPathy_ReturnsTrueOutValue()
+
+            var envars = new TestEnvironment();
+            var git = new TestGit();
+            git.Configuration.Global[$"{section}.{property}"] = new[] {"af_unix:./test.uds"};
```
the "af_unix:" prefix is Linux/MacOS specific.
Windows will have a "//./pipe/" prefix (and/or the backslash version of that string).

jeffhostetler commented on src/shared/Core/ApplicationBase.cs:
```diff
@@ -75,6 +77,31 @@ public Task<int> RunAsync(string[] args)
+            if (Context.Settings.GetTrace2TracingEnabled(out string trace2Value))
+            {
+                var connectionPath = PlatformUtils.IsWindows() ? trace2Value.TrimUntilLastIndexOf("/") :
+                    trace2Value.TrimUntilIndexOf(":");
```
On Windows, we should strip off the UNC host and drive -- so basically strip off the "//./pipe/" prefix.  In this context the "." is the UNC hostname and "pipe" is the mount point -- since all names pipes are created on the special named pipe file system (NPFS) pseudo-drive.  Named pipes can have more slashes (but usually don't).
It's the prefix that is the key.

On Unix, we want to strip off the "af_unix:", "af_unix:stream:" or "af_unix:dgram:" prefix.  (The optional stream or dgram term specifies TCP or UDP and if omitted we should try both types of connections.). (I'm not sure the GOLANG daemon supports both protocols right now, but we should be able to parse the it here.)

ldennington replied:
Thanks for bringing these additional use cases to my attention. It's a perfect candidate for a unit test, let me put some more thought into how we can implement that.

jeffhostetler commented on src/shared/Core/ApplicationBase.cs:
```diff
@@ -75,6 +77,31 @@ public Task<int> RunAsync(string[] args)
+                if (!string.IsNullOrEmpty(connectionPath))
+                {
+                    Context.Trace2.AddListener(new CollectorWriter(
+                            () => new NamedPipeClientStream(".", connectionPath, PipeDirection.Out,
```
I'm confused.  Does NamedPipeClientStream() talk to Unix domain sockets on Unix??
I would have expected this section to be guarded with platform selectors.

mjcheetham replied:
The implementation of the `NamedPipe*Stream` classes on Unix is always UDS, yes. They are also limited to 'byte' mode.

jeffhostetler replied:
ok thanks! i'll keep that in mind. i think we're ok if i always do readline() rather than assuming i always get a whole message.

jeffhostetler commented on src/shared/Core/ApplicationBase.cs:
```diff
@@ -75,6 +77,31 @@ public Task<int> RunAsync(string[] args)
+
+                    if (AssemblyUtils.TryGetAssemblyVersion(out string version))
+                    {
+                        Context.Trace2.WriteVersion(version);
```
The version event is required.  It sets the absolute wall-clock of the start of the program and causes other things to be initialized in the daemon.

Could we change TryGetAssemblyVersion() always return a version number value -- even if something like "0.0" or "0.0.0".

ldennington replied:
I'm hesitant to do this, since other code relies on `TryGetAssemblyVersion` working as it does. Perhaps we could check to see if the result is false and then add a "dummy" version here based on that. @mjcheetham - any thoughts? I'm definitely open to a cleaner solution.

jeffhostetler replied:
That's fine.  Let it get the real version number from the assembly (which will probably be updated by the CI/Release build scripts), and fallback to a constant that you hard-code here if the assembly function fails.

jeffhostetler replied:
(It's really not that big a deal.)

jeffhostetler commented on src/shared/Core/CollectorWriter.cs:
```diff
@@ -0,0 +1,172 @@
+            {
+                // If we're in byte/stream transmission mode rather than message mode
+                // we should signal the end of each message with a line-feed (LF) character.
+                if (_pipeClient.TransmissionMode == PipeTransmissionMode.Byte)
```
We should always write in message mode, so that the daemon doesn't get fragmented messages.

The message should always have a trailing LF (makes debugging and logging easier).

jeffhostetler replied:
Sounds like on the Unix side, we'll always have UDP in byte mode, so ignore this.  The important thing is to always end the message with a LF.  (The JSON encoder should escape/encode  (JSON-quote) any LFs in the actual data.)

jeffhostetler commented on src/shared/Core/CollectorWriter.cs:
```diff
@@ -0,0 +1,172 @@
+                // mechanism and drop this message. We will try to recreate/connect the pipe on the next message.
+                _pipeClient.Dispose();
+                _pipeClient = null;
+                RaiseStateChanged(CollectorWriterState.Failing, ex);
```
If you get a SIGPIPE or something, just give up and turn off telemetry.  Don't try to restart it in the middle of the process.

GCM is only running for the duration of a single command, a missed message in the middle of the lifetime of this one invocation won't affect future invocations.  And the daemon cannot re-sync after an arbitrary dropped message -- for example, if you drop a region-enter, the daemon won't know what to do with the later (not dropped) region-leave.

If git.exe gets an error writing to the socket/pipe, it just turns off Trace2 and silently continues.
It's more important to not slow down the command by waiting to retry (or trying lots of re-connects).

If the daemon drops the connection (or sees that the client has dropped it), it is going to cleanup/flush the incomplete state for this client invocation anyway -- it won't try waiting around for the client to re-connect.

mjcheetham replied:
I would just remove all of the state monitoring and events, and replace it with a simple "failed" flag that is set if the initial `Start/Connect` didn't work, or whenever an exception is thrown sending a message.

Maybe we should log to `Trace1` on failure? :)

jeffhostetler commented on src/shared/Core/CommandContext.cs:
```diff
@@ -22,6 +23,11 @@ public interface ICommandContext : IDisposable
+        /// <summary>
+        /// Application execution timer.
+        /// </summary>
+        Stopwatch ApplicationTimer { get; }
```
If this global timer is being added for Trace2 reasons, I'd add Trace2 to the name of the variable to make it more obvious.

mjcheetham replied:
Feels like `Trace2` should own the timer. In this case, `Stopwatch` should either be created inside of `Main` and passed to the `CommandContext.ctor` (which passes it through to `Trace2.ctor`), or `Trace2` should create its own `Stopwatch`, and `Trace2` should be created directly inside `Main` and passed to `CommandContext.ctor`.

mjcheetham replied:
Or.. why not just store the `DateTime` or `long` Unix timestamp of when the application started, and compute elapsed times as we process events?

`DateTime.Now - DateTime.UnixEpoch`

or

`DateTimeOffset.Now.ToUnixTimeSeconds()`

mjcheetham replied:
BTW.. does Git use local time, or UTC?

jeffhostetler replied:
UTC

jeffhostetler replied:
Each event record needs a "time" field with the current time (UTC with hopefully microsecond resolution).

Some event records add a "t_abs" field of elapsed seconds (float with 6 digits behind the decimal point).
(Yes, this is redundant, but it was helpful to have both in the format.)

Either way works, passing `DateTime.NowUtC()` (or whatever the c# class/method is called) on
the `Write<event>()` call is more direct and has less latency.  Letting the trace helper thread do the
call to NowUTC() is easier, but less accurate since a small amount of time has elapsed because of the
message queuing and background thread handoff.  (But it's not that bad either way.)

Maybe have a "trace2 init / ctor" call in/near Main() that sets a private start-time in the class (and sets
the SID) -- and maybe hides both from the rest of the program.

jeffhostetler commented on src/shared/Core/Trace2.cs:
```diff
@@ -0,0 +1,278 @@
+    }
+
+    public string GetSid()
+    {
```
If I'm reading this right, this will create a new SID every time it is called.  And it'll keep getting longer (because of the setenv at the bottom will cause the getenv on the next call to get the updated value).

Shouldn't this have a static variable hidden within it and just return it for subsequent calls?

ldennington replied:
Good catch, thank you!

jeffhostetler commented on src/shared/Core/Trace2.cs:
```diff
@@ -0,0 +1,278 @@
+            File = filePath,
+            Line = lineNumber,
+            Code = code,
+            // In Git, the "fmt" field is used to write exception information minus any PII.
```
I think this comment is left over from a cut/paste.

ldennington replied:
On further examination, this appears to be _both_ an accidental cut 'n paste and an unfinished thought 😆 .

jeffhostetler commented on src/shared/Core/Trace2.cs:
```diff
@@ -0,0 +1,278 @@
+        var sidSuffix = Guid.NewGuid();
+        var currentSid = $"{parentSid}{sidSuffix}";
+        
+        _context.Environment.SetEnvironmentVariable(CurrentSidVariable, currentSid);
```
Oh, I see, you're writing out a new GCM_...SID rather that overwriting the inherited GIT_...SID envvar.
That's why it doesn't keep getting longer with each call.

Instead, we should overwrite the GIT_...SID envvar so that when GCM invoked `git config`, the child will inherit the modified SID.  This will put GCM in the trace-graph between the `git fetch` and the `git config`, for example.

mjcheetham replied:
I wonder if we should just go all in on `GIT_TRACE2` and not introduce `GCM_` specific variables.

jeffhostetler commented:
This looks like a good start.  Thanks for all your work on this.

Arezoomohammadalizadeh commented:
در سه‌شنبه 24 ژانویه 2023 در 2:47 Lessley Dennington <
***@***.***> نوشت:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/shared/Core/Trace2.cs
> <https://github.com/GitCredentialManager/git-credential-manager/pull/1045#discussion_r1084669642>
> :
>
> > +    /// <summary>
> +    /// Add a listener to the trace writer.
> +    /// </summary>
> +    /// <param name="listener">The listener to add.</param>
> +    void AddListener(CollectorWriter listener);
>
> Yes, the CollectorWriter that is added here sets up the queue that
> receives TRACE2 messages and eventually writes them to the daemon. Perhaps
> AddWriter would be a more appropriate name...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/GitCredentialManager/git-credential-manager/pull/1045#discussion_r1084669642>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AWWGRIDNA5ZPWNJBLLNBGYLWT4GQ3ANCNFSM6AAAAAAUCAPOJU>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID:
> <GitCredentialManager/git-credential-manager/pull/1045/review/1266579123@
> github.com>
>
-- 
null

mjcheetham commented on src/shared/Core/Trace2.cs:
```diff
@@ -0,0 +1,93 @@
+
+namespace GitCredentialManager;
+
+// The different event types tracked in the TRACE2 tracing
+// system.
```
Make this a `///` doc comment.

mjcheetham commented on src/shared/Core/Trace2.cs:
```diff
@@ -0,0 +1,93 @@
+    void AddListener(CollectorWriter listener);
+}
+
+public class Trace2: DisposableObject, ITrace2
```
nit

```diff
-public class Trace2: DisposableObject, ITrace2
+public class Trace2 : DisposableObject, ITrace2
```

mjcheetham commented on src/shared/Core/Trace2.cs:
```diff
@@ -0,0 +1,93 @@
+        _writer.Start();
+    }
+
+    public new void Dispose()
+    {
+        if (_isDisposed)
+        {
+            // This instance has already been disposed
+            return;
+        }
+        
+        _writer.Dispose();
+        _isDisposed = true;
+    }
```
If we're inheriting from `DisposableObject` we shouldn't be hiding the base `Dispose` method (using `new`).

`DisposableObject` implements the `IDisposable` pattern and makes it easier to implement by offering two extension points:

```csharp
/// <summary>
/// Called when unmanaged resources should be released and memory freed.
/// </summary>
protected virtual void ReleaseUnmanagedResources() { }

/// <summary>
/// Called when managed resources should be released.
/// </summary>
protected virtual void ReleaseManagedResources() { }
```

Here you'd just want to `override` `ReleaseManagedResources` to dispose of `_writer`.


```csharp
protected override void ReleaseManagedResources()
{
    _writer.Dispose();
    base.ReleaseManagedResources();
}
```

mjcheetham commented on src/shared/Core/Trace2.cs:
```diff
@@ -0,0 +1,93 @@
+
+public abstract class CollectorMessage
+{
+    [JsonProperty("event")] public Event Event { get; set; }
+    [JsonProperty("sid")] public string Sid { get; set; }
+    [JsonProperty("thread")] public string Thread { get; set; }
+    [JsonProperty("time")] public string Time { get; set; }
+    [JsonProperty("file")] public string File { get; set; }
+    [JsonProperty("line")] public double Line { get; set; }
```
I know many default IDE formatting rules put these attributes one the same line as the property, but  think it's better to have them in the more traditional style:


```csharp
[JsonProperty("event")]
public Event Event { get; set; }
```

ldennington replied:
Sure, that's an easy change, and I like the look of it better as well.

ldennington replied:
Easier to stack additional attributes if we need to as well.

mjcheetham commented on src/shared/Core/CommandContext.cs:
```diff
@@ -92,7 +92,7 @@ public CommandContext(string appPath, string installDir)
             Streams = new StandardStreams();
             Trace   = new Trace();
-            Trace2  = new Trace2();
+            Trace2  = new Trace2(this);
```
At first view I raised my eyebrow about passing the entire `ICommandContext` to any component that hangs off of `CommandContext` itself, but then noticed that we actually do have one component that does this: `CredentialStore`.

I think we should avoid this pattern unless we're creating `Trace2` as the last thing in the `CommandContext` constructor.

Imagine if, inside of `Trace2.ctor` we try and access a service or component on `ICommandContext` that hasn't been setup yet?

I think `Trace2` only needs `IEnvironment` right now.

mjcheetham commented on src/shared/Core/EnvironmentBase.cs:
```diff
@@ -141,6 +150,9 @@ internal virtual bool TryLocateExecutable(string program, ICollection<string> pa
             return false;
         }
+
+        public abstract void SetEnvironmentVariable(string variable, string value,
+            EnvironmentVariableTarget target = EnvironmentVariableTarget.Process);
```
Both of the implementations of `SetEnvironmentVariable` (POSIX and Windows) are the same. We should just move the implementation to this `EnvironmentBase` class instead.

Will need to move `GetCurrentVariables` to a `protected abstract` method on `EnvironmentBase`, and then update the constructors that are calling this statically and replace the implementation with a `null -> GetCurrentVariables()` call.

E.g.

```diff
-public WindowsEnvironment(IFileSystem fileSystem)
-    : this(fileSystem, GetCurrentVariables()) { }
-
-internal WindowsEnvironment(IFileSystem fileSystem, IReadOnlyDictionary<string, string> variables)
+internal WindowsEnvironment(IFileSystem fileSystem, IReadOnlyDictionary<string, string> variables = null)
    : base(fileSystem)
{
-    EnsureArgument.NotNull(variables, nameof(variables));
-    Variables = variables;
+    Variables = variables ?? GetCurrentVariables();
}
```

mjcheetham commented on src/shared/Core/Trace2.cs:
```diff
@@ -32,6 +42,28 @@ public class Trace2: DisposableObject, ITrace2
+
+    public string GetSid()
+    {
+        var parentSid = "";
+        if (_context.Environment.Variables.ContainsKey(ParentSidVariable))
+        {
+            parentSid = _context.Environment.Variables.FirstOrDefault(name => 
+                name.Key.Equals(ParentSidVariable, StringComparison.OrdinalIgnoreCase)).Value + "/";
+        }
+
+        var sidSuffix = Guid.NewGuid();
+        var currentSid = $"{parentSid}{sidSuffix}";
```
This could be simplified using `TryGetValue` on the `IReadOnlyDictionary<,>` and `string.Join` to join the SIDs:

```csharp
var sids = new List<string>();
if (Variables.TryGetValue(ParentSidVariable, out string parentSid))
{
    sids.Add(parentSid);
}

string childSid = Guid.NewGuid().ToString("D");
string combinedSid = string.Join('/', sids);
```

Note that we should be explicit when converting a `Guid` to a `string` representation. What sort of UUID format does Git use for it's SID (if it even uses a UUID)?

<img width="702" alt="image" src="https://user-images.githubusercontent.com/5658207/214365591-1fd76ca7-f2d1-4af7-9ea3-18e12b2e8dd1.png">

jeffhostetler replied:
Git uses a custom-defined unique string (like a GUID/UUID, but without adding another third-party library to the linkage of Git on some platforms who will not be named).

I suggested a Guid here because you have the class in the .Net runtime and get it for free.

The format doesn't matter really. I just need enough bits to have some entropy.
Let's go with "D" format.  The curly braces will cause JSON-quoting issues.

jeffhostetler replied:
I'm not sure in your suggestion where the new childSid gets added to the result.

Also, we want to make sure we only have '/' between the components.  And not a leading '/'.

mjcheetham commented on src/shared/Core/Trace2.cs:
```diff
@@ -32,6 +42,28 @@ public class Trace2: DisposableObject, ITrace2
+        _context = context;
+    }
+
+    public string GetSid()
```
I'd say avoid making this public if it modifies the environment _and_ returns the new SID. Either the caller should `Get` the new SID and update the environment itself, or this method shouldn't be called `Get`, but something like `Set` or `Update`.

mjcheetham commented on src/shared/TestInfrastructure/Objects/NullTrace.cs:
```diff
@@ -47,4 +47,20 @@ bool ITrace.IsSecretTracingEnabled
+        #region ITrace2
+        void ITrace2.AddListener(CollectorWriter listener) { }
+
+        public string GetSid() { return ""; }
```
`return string.Empty`

mjcheetham commented on src/shared/Core/ApplicationBase.cs:
```diff
@@ -75,6 +77,31 @@ public Task<int> RunAsync(string[] args)
+                        Context.Trace2.WriteVersion(version);
+                    }
+
+                    var argv = new List<string>() { Context.ApplicationPath };
+                    argv.AddRange(args);
+                    Context.Trace2.WriteStart(argv, Context.ApplicationTimer.ElapsedMilliseconds);
```
We've really lost the actual `argv` here since we don't pass it down beyond `Main`.

I wonder if the we need to pass these `Trace2` vars to the `CommandContext` inside of `Main`, along with the constructed timer.

jeffhostetler replied:
you could also fake it here.  just pass something like ["gcm", "store"].
I don't care so much about the pathname to the exe or the other helper args.
we just need to be able to distinguish the different operations.

if you can get the real command line, great, but it's not super critical.

mjcheetham commented on src/shared/Core/ApplicationBase.cs:
```diff
@@ -75,6 +77,31 @@ public Task<int> RunAsync(string[] args)
+
+                    var argv = new List<string>() { Context.ApplicationPath };
+                    argv.AddRange(args);
+                    Context.Trace2.WriteStart(argv, Context.ApplicationTimer.ElapsedMilliseconds);
```
I assume the application timer get used to timestamp all events based on the current `ElapsedMilliseconds` since app start?

Why don't we just store/accept a `DateTime` (or `DateTimeOffset`, or UNIX 64-bit timestamp `long`) that is computed once in `Main` and passed through to `CommandContext` and `Trace2`?

Elapsed times can be computed instead of maintaining a running `Stopwatch`.

jeffhostetler replied:
Yeah, just get the current utc time in Main() and let each event generator do a Now() to emit in the record and let that routine do the subtraction to get the `t_abs` field for the record. (Yes, this is a redundancy in the format.)

Later, we'll need to talk about region-relative stuff and they have a hierarchy of clocks for nested regions.  I can explain later.

ldennington replied:
Ignore my previous question on this topic - thanks for answering here!

mjcheetham commented on src/shared/Git-Credential-Manager.UI.Avalonia/Program.cs:
```diff
@@ -44,7 +45,8 @@ private static void AppMain(object o)
             string appPath = ApplicationBase.GetEntryApplicationPath();
             string installDir = ApplicationBase.GetInstallationDirectory();
-            using (var context = new CommandContext(appPath, installDir))
+            // Pass in a "dummy" stopwatch, since TRACE2 is not enabled for helper exes.
```
Any reason why not to implement Trace2 for helper UI? Or is this a 'todo' in the future thing?

ldennington replied:
My reasoning was that I didn't see a lot of `TRACE` stuff in the helper UI code. Could definitely be a future thing though.

mjcheetham reviewed (commented):
Looking good so far! I have some comments about simplifying the collector/queue, going all in on `GIT_` variables, and how we're instantiating `Trace2` and the stopwatch.

jeffhostetler commented:
I just responded to all of Matthew's questions that I had an opinion on.

ldennington commented:
Changes in v2:

- CollectorWriter now inherits from `DisposableObject`.
- Removed the GCM-specific `TRACE2` and `TRACE2_SID` environment variables in favor of piggybacking off Git’s existing variables.
- Updated comments on AddListener to better describe its purpose.
- Removed `Stopwatch()` in favor of calculating elapsed times using Unix timestamps. Additionally, updated times that were using `DateTime.Now` to `DateTime.UtcNow`.
- Ensured event target parsing is correct for both Posix and Windows systems (using `TrimUntilLastIndexOf` and moved the logic into a new `TryParseEventTarget()` method in `Trace2.cs`. Handling for other event targets will happen in future PRs with their respective implementations.
- Ensured there is always a version (even if it’s just a placeholder) to send to `WriteVersion()`.
- Ensured messages sent to the telemetry daemon/OTel collector always end with a line feed.
- Ensured we’re making a call to `WriteExit()`.
- Fixed incorrect implementation of DisposableObject in `Trace2.cs` and `CollectorWriter.cs`.
- Removed state from `CollectorWriter` in favor of a `Failed` flag.
- Condensed Sid logic - updated method name from `GetSid()` to `SetSid()` and moved call to `Trace2` constructor.
- Ensured we are writing the correct args in `WriteStart()`.

jeffhostetler commented on src/shared/Core.Tests/SettingsTests.cs:
```diff
@@ -280,6 +279,86 @@ public void Settings_IsTracingEnabled_EnvarPathy_ReturnsTrueOutValue()
+            }
+            else
+            {
+                git.Configuration.Global[$"{section}.{property}"] = new[] {"//./pipe/git-c780ac06-135a-4e9e-ab6c-d41e2d265baa"};
```
Both "af_unix:./test.uds" and this "//./pipe/git-c78a...." string literals should not be in the source.  (They're not secrets or anything.)   However, they refer to whatever socket/pipe that we have the/a server listening on.

Is the goal here to actually use the pathnames and connect -- or just to confirm that they are non-empty (from an earlier test setup) (or return these values so that another test can confirm the correct value is returned) -- either way, I think we could use some fictional literals or something.

ldennington replied:
The goal of the test is to show that if someone has the `trace2.eventtarget` config set, GCM will correctly detect that tracing is enabled. But totally fine with putting some dummy values in there.

ldennington replied:
That's much better anyway because it removes the weird platform switch.

jeffhostetler commented on src/shared/Core/Trace2.cs:
```diff
@@ -0,0 +1,316 @@
+    /// </summary>
+    /// <param name="eventTarget">
+    /// The TRACE2 event target value. Supports the following formats:
+    /// * af_unix:<socket_type>:<absolute-pathname> -  writes to Unix Domain Socket
```
sorry to nit, this should be "af_unix:[<socket_type>:]<absolute-pathname>"

that is, the socket type is optional.  however, i generally don't set it anymore. (I had a reason at the time.)
If the C# NamedPipeClientStream class only does one type (I forget which), then we might document that.
I'm not sure we need to explicitly throw an error if the eventTarget string contains the other type -- i'm sure
that the network stack will let us know....

jeffhostetler commented on src/shared/Core/Trace2.cs:
```diff
@@ -0,0 +1,316 @@
+    /// <param name="eventTarget">
+    /// The TRACE2 event target value. Supports the following formats:
+    /// * af_unix:<socket_type>:<absolute-pathname> -  writes to Unix Domain Socket
+    /// * \\.\pipe\git-<GUID> - writes to Windows Named Pipe
```
sorry to nit again, this value just needs to be a path to a named pipe on the local host, so "//./pipe/<anything>" is sufficient.

Sorry if I confused you earlier.  The "git-" and <guid> literals were just the values I was using in the existing daemon.

ldennington replied:
No need to apologize - I'm glad you caught this!

jeffhostetler commented on src/shared/Core/Trace2.cs:
```diff
@@ -0,0 +1,316 @@
+    /// Writes exit code and execution time to trace writer.
+    /// </summary>
+    /// <param name="code">Application exit code.</param>
+    /// <param name="timeInMilliseconds">Elapsed time in milliseconds since GCM started.</param>
```
timeInMilliseconds is no longer in the method args.

jeffhostetler commented on src/shared/Core/Trace2.cs:
```diff
@@ -0,0 +1,316 @@
+    
+    public long ApplicationStartTime { get; }
+
+    public Trace2(IEnvironment environment, string[] args, long applicationStartTime = 0)
```
(_I'm having a hard time jumping from caller to callee in the PR view, so apologies in advance._)
So I'm assuming that this ctor() is called directly/indirectly from Main().

Wouldn't it be better to have the Main() pass down a DateTime set to UtcNow (as observed in Main())
or just let the ctor() compute it.  In either case, store that absolute_start_time as a private field inside the Trace2 class.

Then have context.Trace2.WriteVersion() just used the stored absolute start time in the message.

Then have the event methods (context.Trace2.WriteExit() and context.Trace2.WriteStart()) NOT take an elapsed time argument, but rather have them call DateTime.UtcNow() and use that in the JSON Time field.

For the events that need to emit a "t_abs" / elapsed time field, let those WriteFOO() methods do the subtraction and hide all of this from the callers.

ldennington replied:
Sure, that's fine. I went ahead and made the field in `Trace2.cs` private, set the time in `WriteVersion` to the value of the field, and updated the WriteFOO() methods to do the subtraction.

jeffhostetler commented on src/shared/Git-Credential-Manager/Program.cs:
```diff
@@ -11,9 +12,12 @@ public static class Program
         public static void Main(string[] args)
         {
+            // Set up timer for TRACE2
+            var applicationStartTime = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
```
I just made a comment in Trace2.c about moving all of the time-math into it, but I wanted to say that we should be careful converting UTC times to seconds -- we're measuring mili- and micro-second times and here we're lopping of the time to integer seconds.  There is a To...Milliseconds() method that would be better (inside the WriteFOO() events).

That is assuming that there isn't a way to have the WriteFOO() methods subtract two DateTime values directly and get an elapsed TimeSpan expressed in a microsecond value and then format that as a time-in-seconds floating point value in the JSON event.

ldennington replied:
I think I've got an extension method that's doing what we want here. Let me know what you think.

ldennington commented:
Changes since v2:

1. Removed local `pipe` variable from `CollectorWriter` `Start()` method.
2. Removed `af_unix:./test.uds` and `//./pipe/` string literals from `IsTrace2TracingEnabled` test.
3. Updated all times to microseconds (with the help of a new `DateTimeOffset` extension `ToUnixTimeMicroseconds`).
4. Calculated elapsed time for `Start` and `Exit` events in the methods that write these events (rather than calculating in the caller(s)).

mjcheetham commented on src/shared/Core/ITrace2Writer.cs:
```diff
@@ -0,0 +1,10 @@
+
+public interface ITrace2Writer : IDisposable
+{
+    public bool Failed { get; }
+
+    public void Write(Trace2Message message);
```
No need for the `public` access modifier in an `interface`.

```suggestion
   bool Failed { get; }

   void Write(Trace2Message message);
```

mjcheetham commented on src/shared/Core/DateTimeOffsetExtensions.cs:
```diff
@@ -0,0 +1,16 @@
+
+public static class DateTimeOffsetExtensions
+{
+    private static readonly DateTimeOffset UnixEpoch =
+        new DateTimeOffset(1970, 1, 1, 0, 0, 0, TimeSpan.Zero);
```
I was about to comment that `DateTimeOffset.UnixEpoch` already exists, but of course it does not in .NET Framework... 😢

mjcheetham commented on src/shared/Core/CommandContext.cs:
```diff
@@ -78,7 +85,7 @@ public interface ICommandContext : IDisposable
     public class CommandContext : DisposableObject, ICommandContext
     {
-        public CommandContext(string appPath, string installDir)
+        public CommandContext(string appPath, string installDir, string[] argv, long applicationStartTime = 0)
```
Defaulting to `0` feels wrong. Why don't we default to computing the current time at the point the constructor is called here?

Given the first thing that almost all of the Program.cs entry points do is `new CommandContext(...)` it would be neater.

```csharp
// Program.cs
public static void Main(string args[])
{
    var appPath = ApplicationBase.GetApplicationPath();
    using (var context = new CommandContext(appPath, args))
    {
        ...
    }
}

// CommandContext.cs
public CommandContext(string appPath, string[] argv, long? startTime = null)
{
    long appStartTime = startTime ?? DateTimeOffset.UtcNow.ToUnixTimeMicroseconds();

    // The rest of CommandContext setup
    ...
}
```

We could also move the `ApplicationBase.GetApplicationPath` call in to the constructor too, if `null` was passed.

```csharp
// Program.cs
public static void Main(string args[])
{
    using (var context = new CommandContext(args))
    {
        ...
    }
}

// CommandContext.cs
public CommandContext(string[] argv, string appPath = null, long? startTime = null)
{
    long appStartTime = startTime ?? DateTimeOffset.UtcNow.ToUnixTimeMicroseconds();
    ApplicationPath = appPath ?? GetApplicationPath();

    // The rest of CommandContext setup
    ...
}
```

The ability to pass params other than `argv` to the CommandContext would be left if we wanted to test using a real `CommandContext` (rather than the `TestCommandContext`) but with stable paths and timestamps.

mjcheetham reviewed (commented):
A few comments I have so far

ldennington commented:
Changes since v3:

1. Re-architected solution to more easily extend to additional event format targets and outputs (stderr/file). I have diagrammed the main components of the new solution in the hopes it’ll make it easier for reviewers to parse this PR (recommended to open in a separate window for viewing):
![trace2](https://user-images.githubusercontent.com/11321782/217341688-d54c1775-04ee-48e9-a52f-dbcc3998c14f.png)
3. Added support for the normal format target.
4. Refactored shared `Trace` logic into a new `TraceUtils` class.
5. Re-organized calling of current events (`Version`, `Start`, and `Exit`) to occur via `Start` and `Stop` methods.
6. Updated application start time to be a `DateTimeOffset` rather than a `long`. This removed the need for a `DateTimeOffsetExtensions` class to convert to and from Unix time stamps.
7. Formatting updates to correctly write all events to different event format targets and outputs:
    - Writing all file names and paths in lower case.
    - Ensuring the `Time` field is the time at which the event is being written (rather than the application start time).
    - Writing in local time for the normal format target and UTC time for the event format target.
    - Ensuring line numbers are written as integers (rather than doubles).

derrickstolee commented on src/shared/Core/Trace2.cs:
```diff
@@ -48,6 +59,7 @@ protected override void ReleaseManagedResources()
         base.ReleaseManagedResources();
     }
 
+<<<<<<< HEAD
```
Looks like an errant merge conflict. Maybe you take care of it in a later commit.

ldennington replied:
I should have reviewed the commits more carefully after my last rebase. Thank you for catching.

derrickstolee commented on src/shared/Core/Trace2.cs:
```diff
@@ -59,21 +91,6 @@ protected override void ReleaseManagedResources()
         base.ReleaseManagedResources();
     }
 
-<<<<<<< HEAD
```
Here's the deletion.

derrickstolee commented on src/shared/Git-Credential-Manager/Program.cs:
```diff
@@ -75,7 +74,7 @@ public static void Main(string[] args)
                                   .GetAwaiter()
                                   .GetResult();
-
+                
```
Do you have a setting in your IDE to trim trailing whitespace on save? Could help to avoid lines like this.

derrickstolee replied:
`git rebase --whitespace=fix` should fix it, too.

ldennington replied:
Looks like Rider has a [nice feature for this](https://youtrack.jetbrains.com/issue/RIDER-9687/Add-option-to-trim-trailing-whitespace-on-save#focus=Comments-27-2437597.0-0). Thanks!

derrickstolee reviewed (approved):
I took a quick look through, and only had some nitpicks along the way.

However, one thing I was looking forward to seeing was a test that could demonstrate that the `git-credential-manager` executable would output the trace2 output (in event or normal form) as much as possible. The event form creates a JSON event on a per-line basis, so being able to parse that JSON even and checking that it has the necessary items would be helpful. (Any issues with _extra_ items?)

derrickstolee reviewed (approved):
:shipit: Looks like a strong foundation and anything from here on out can be forward-fixed as we get more use out of it (and an end-to-end testing framework).

jeffhostetler commented:
You merged it as I was about to push the approve button.... :-)
This looks good. Nice work.
This will give us a chance to kick the tires now. Thanks.
